python/ipywidgets/ipywidgets/widgets/utils.py

```diff
@@ -0,0 +1,35 @@
+# Copyright (c) Jupyter Development Team.
+# Distributed under the terms of the Modified BSD License.
+
+import sys
+import warnings
+
+# This function is from https://github.com/python/cpython/issues/67998
+# (https://bugs.python.org/file39550/deprecated_module_stacklevel.diff) and
+# calculates the appropriate stacklevel for deprecations to target the
+# deprecation for the caller, no matter how many internal stack frames we have
+# added in the process. For example, with the deprecation warning in the
+# __init__ below, the appropriate stacklevel will change depending on how deep
+# the inheritance hierarchy is.
+def external_stacklevel(internal):
+    """Find the first frame that doesn't any of the given internal strings
+
+    The depth will be 2 at minimum in order to start checking at the caller of
```

Comment:
This logic seems to be using the assumption that `external_stacklevel` is always called from `deprecation`. If this is intended, we might as well make it private (`_external_stacklevel`).

Reply:
I thought this default made sense for general use:

level 0: inside external_stacklevel
level 1: the line calling external_stacklevel (usually the warnings.warn line)
level 2: whoever called the function that is trying to give out a deprecation warning

In general, I think it makes sense to point to the code that is calling the function giving the deprecation warning. If I was always calling this only from the deprecate function below, I would have put the default at level 3.

Reply:
To clarify some points:
```python
import ipywidgets
def my_function():
    # what should this output?
    print(ipywidgets.widgets.utils.external_stacklevel([]))
    # and what about this?
    print(ipywidgets.widgets.utils.external_stacklevel(['ipywidgets']))
my_function()
```
This currently prints `2` and `2`, which I argue does not make sense when used directly, when you consider the docstring "Find the first frame that doesn't any of the given internal strings". If that description was correct, I would expect `0` and `1`.

Reply:
If the second sentence of the docstring is taken into account, that explains the behavior of the function, but that means that you would never really know what stack level you were at inside `my_function`.

Reply:
as such, my comment is that the current behavior only makes sense when the stack level of the calling function of `external_stacklevel` is never of interest (which will be the case inside `deprecation`, but does not make sense as a general export that is available from non `_private` paths from the root import of `ipywidgets`.

Reply:
Ah, I see my mistake - I should adjust the number I'm giving back, since it will be used at a different level than what is was calculated on.